### PR TITLE
feat: Add test tools docs

### PR DESCRIPTION
**Pull request by hampuslavin**
## Descrption / reasoning
The idea is to divide into different pages that has different focus areas. Currently I have the following in mind:
- **Get started**: Easily digested get started info, minimum for getting the simplest test to run
- **The basics**: Additional documentation you will absolutely need to know within "the first hour" of writing tests
- **Advanced exmples**: More complex test setup examples. Can still be common cases but not something you necessarily need to write a first test suite.
- **Best practises**: Inspired by Cypress' (e2e web test framework) [best practices page](https://docs.cypress.io/guides/references/best-practices), as inevitably there will always be cases where a user can do a thing in multiple ways. I've found the Cypress page to be tremendously helpful over the years. It also helps to educate and onboard team members who are new to a framework to quickly get into the "philosophy" of the tool.
image

❓ TODO / Question for reviewer:
- Consider what more best practises to add. Maybe some general testing principles that applied outside of Serverpod as well?

### CODE REVIEW

File: docs/06-concepts/18-testing.md
Comment (christerswahn) [inline docs/06-concepts/18-testing.md:12]: Since this quote starts with an import it gives the impression it is a complete file. Suggest you make it complete with main() etc as well:

```
import 'package:test/test.dart';
import '../../test_tools/serverpod_test_tools.dart';

void main() {
  withServerpod('Given Example endpoint', (endpoints, session) {
    test('when calling `hello` then should return greeting', () async {
      final greeting = await endpoints.example.hello(session, 'Michael');
      expect(greeting, 'Hello, Michael!');
    });
  });
}

```

Comment (hampuslavin) [reply docs/06-concepts/18-testing.md:12]: You're completely right, missed that! 👍

Comment (hampuslavin) [reply docs/06-concepts/18-testing.md:12]: Fixed!

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [inline docs/06-concepts/18-testing/01-get-started.md:11]: Suggestion: Also link to the complete file after the modifications.

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [inline docs/06-concepts/18-testing/01-get-started.md:121]: Suggestion(NIT): Use another word than "world". Maybe "server state", "session state" or "server and session state"?

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [inline docs/06-concepts/18-testing/01-get-started.md:122]: Question: Does this also contain your module endpoints?

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [inline docs/06-concepts/18-testing/01-get-started.md:138]: Suggestion: End the tutorial by running the test also.

Since we are using a different path for our tests we might as well include that in the tutorial as well.

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [inline docs/06-concepts/18-testing/01-get-started.md:98]: Question: Should we include a small motivation for why we don't generate this inside of the normal "test" folder? The reason for separating the integration tests from the unit tests.

Comment (SandPod) [reply docs/06-concepts/18-testing/01-get-started.md:98]: Maybe that is more for the best practices section?

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:31]: Suggestion/Discussion: Don't expose the implementation details of the TestSessionBuilder in the docs.

The implementation details for the TestSessionBuilder are going to evolve over time. Will we keep up with the documentation? Maybe we can explain it more abstract and then go into single details for each "copyWith" capability instead?

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:48]: Suggestion(NIT): Can you force these to be on the same line?

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:98]: Question: Is the authentication important for this example or could it be removed? Seems like redundant information to prove the seeding point.

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:146]: Suggestion: Should we present these in a table instead?

Feels like that is easier on the eye to parse and provide an overview?

Comment (SandPod) [reply docs/06-concepts/18-testing/02-the-basics.md:146]: We have used tables in other sections of our documentation. Might be worth keeping this consistent.

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:224]: Suggestion: Use a table for the exceptions.

Same feedback here, a table gives a clear overview of the options and then we don't need to remember to sync the implementation details with the copied code here.

File: docs/06-concepts/18-testing/04-best-practises.md
Comment (SandPod) [inline docs/06-concepts/18-testing/04-best-practises.md:23]: Suggestion: Mention the configuration option that changes this behavior, if someone has not read that section of the documentation it can be good to point them there.

**Review by SandPod — COMMENTED** [review]
Awesome work!

I like the structure presented here, have a couple of comments for different sections.
Ping me if anything is unclear!

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:121]: Yes, great point!

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:98]: Hmmm, great point. Could be both 🤷 
image

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:48]: Dart formatter don't want it 😢

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:98]: Great catch, we should absolutely remove that 👍

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:146]: If we use tables in other places it makes sense to be consistent 👍

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:138]: Can you clarify what you mean? Adding `dart test ..` or?

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:122]: Nope, I think the philosophy at least in this first iteration should be that the module should have its own tests (using test tools). If this is too limiting we can add it of course. So that said, I'll add a note about that 👍

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:11]: Do you mean linking to the file in the Serverpod repo?

Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:11]: Then I'll have to wait for the templates PR to be merged first.

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:31]: I would argue that this is not implementation details as I'm only showing the public interface (without any implementation code). However we phrase this part - if the public interface changes that will inevitable require us to sync with the docs.

That said, maybe I can dissect this an lift out the comments as regular documentation text and then have the functions as snippets. E.g. something along these lines: 
>  `copyWith` creates a new unique session with the provided properties. This is useful for setting up different session states in the tests or simulating multiple users.
```dart
  TestSessionBuilder copyWith({
    AuthenticationOverride? authentication,
    bool? enableLogging,
  });
```

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [reply docs/06-concepts/18-testing/01-get-started.md:11]: Yes, or possibly have it as a foldout in the foldout? Just so that it is possible to do a complete copy paste of the file if required.

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [reply docs/06-concepts/18-testing/01-get-started.md:122]: It might make sense to include these in your generated endpoints. Since they essentially are just normal endpoints right?

File: docs/06-concepts/18-testing/01-get-started.md
Comment (SandPod) [reply docs/06-concepts/18-testing/01-get-started.md:138]: Yes, exactly. Since the user can't just run `dart test` to run these tests but need to specify the path to the `integrat_test` folder we should have it as a step.

So `dart test integration_test`. and then maybe in the advanced examples include information about `--concurrency=1` when preserving state in the database between tests.

File: docs/06-concepts/18-testing/04-best-practises.md
Comment (SandPod) [inline docs/06-concepts/18-testing/04-best-practises.md:3]: suggestion: Add a description for the rule.

Does it make sense to always have a description for the rule?

Same for the new `Unit and integration tests {#unit-integration}`. Instead of having the general description in the "Don't" place it at the top.

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [reply docs/06-concepts/18-testing/02-the-basics.md:31]: You are correct, this is the interface, but the comments we have there and the structure of the code is implementation details. So I think it makes more sense to highlight the specific interfaces and keep away from documenting class declarations.

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:214]: Suggestion: Rephrase to more reflect these are the exceptions that can be thrown by the test framework during a method call.

For example: `The following exceptions are used by the test framework....`

Comment (SandPod) [reply docs/06-concepts/18-testing/02-the-basics.md:214]: Should the transaction exception be part of this list as well?

**Review by SandPod — COMMENTED** [review]
Some more minor NITs, do you want to convert this to a PR instead of it being a draft?

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:214]: Since the different exceptions are thrown in different scenarios, I wrote those things on each exception. The transaction case is throwing an `InvalidConfigurationException`. This is a general exception, but it is also documented in the concurrent transacations section that this is what will be thrown.

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:122]: Tracked here, we may or may not wait for its PR before merging in: https://github.com/serverpod/serverpod/issues/2830

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md]: Made small clarification: 49270cc

File: docs/06-concepts/18-testing/01-get-started.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/01-get-started.md:11]: Makes sense!

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:28]: Tried a rework here, but it's a bit tricky. Let me know what you think: 049c08d

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:3]: Suggestion: Shorten to `Set up test scenario`.

Menu looks a bit cramped with this long of a headline.

File: docs/06-concepts/18-testing/04-best-practises.md
Comment (SandPod) [inline docs/06-concepts/18-testing/04-best-practises.md:23]: Suggestion: Change this to only use 2 `#`.

The right side menu on this page is a bit funky, do we really need "do" and "don't" there? maybe we could use 4 `#` to remove them? Or I think there is another way to exclude something from the menu also.

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (SandPod) [inline docs/06-concepts/18-testing/02-the-basics.md:79]: Question: Is the additional `{#seeding-the-database}` really required here?

Comment (SandPod) [reply docs/06-concepts/18-testing/02-the-basics.md:79]: As long as there are no collisions I think this name is automatically created.

If it is not required, please check this for all headers that have this addition.

**Review by SandPod — COMMENTED** [review]
Couple of minor last comments!

Then I think this is good enough to release :)

File: docs/06-concepts/18-testing/02-the-basics.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/02-the-basics.md:79]: Oh, I didn't realize it was auto-generated. Will check all of these 👍

File: docs/06-concepts/18-testing/04-best-practises.md
Comment (hampuslavin) [reply docs/06-concepts/18-testing/04-best-practises.md:23]: I'll check if there's a way to hide them from the menu!

File: docs/06-concepts/18-testing/03-advanced-examples.md
Comment (vlidholt) [inline docs/06-concepts/18-testing/03-advanced-examples.md:10]: I think the that `_` is generally known to be ignored (I'd propose we skip the comment here).

Comment (hampuslavin) [reply docs/06-concepts/18-testing/03-advanced-examples.md:10]: Good point!

File: docs/06-concepts/18-testing/04-best-practises.md
Comment (SandPod) [inline docs/06-concepts/18-testing/04-best-practises.md:3]: Nice! 🙌

**Review by SandPod — APPROVED** [review]
Great work!

I think this looks good to me! :D Let's test it out and see if people understand it :)

**Pull request merged**
